core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java

```diff
@@ -167,19 +167,19 @@ public UnassignedShardDecision makeAllocationDecision(final ShardRouting unassig
                 // let BalancedShardsAllocator take care of allocating this shard
                 logger.debug("[{}][{}]: missing local data, will restore from [{}]",
                              unassignedShard.index(), unassignedShard.id(), unassignedShard.recoverySource());
-                return UnassignedShardDecision.DECISION_NOT_TAKEN;
+                return ShardAllocationDecision.DECISION_NOT_TAKEN;
             } else if (recoverOnAnyNode) {
                 // let BalancedShardsAllocator take care of allocating this shard
                 logger.debug("[{}][{}]: missing local data, recover from any node", unassignedShard.index(), unassignedShard.id());
-                return UnassignedShardDecision.DECISION_NOT_TAKEN;
+                return ShardAllocationDecision.DECISION_NOT_TAKEN;
             } else {
                 // We have a shard that was previously allocated, but we could not find a valid shard copy to allocate the primary.
                 // We could just be waiting for the node that holds the primary to start back up, in which case the allocation for
                 // this shard will be picked up when the node joins and we do another allocation reroute
                 logger.debug("[{}][{}]: not allocating, number_of_allocated_shards_found [{}]",
                              unassignedShard.index(), unassignedShard.id(), nodeShardsResult.allocationsFound);
-                return UnassignedShardDecision.noDecision(AllocationStatus.NO_VALID_SHARD_COPY,
-                    "shard was previously allocated, but no valid shard copy could be found amongst the current nodes in the cluster");
+                return ShardAllocationDecision.no(AllocationStatus.NO_VALID_SHARD_COPY, explain ?
+                    "shard was previously allocated, but no valid shard copy could be found amongst the nodes in the cluster" : null);
```

Comment:
Can you collapse the check into a single line, it's awkward to read

```java
explain ?
"..." : null
```

Since it splits the check and the result, so maybe

```java
return ShardAllocationDecision.no(AllocationStatus.NO_VALID_SHARD_COPY,
        explain ? "shard was ... in the cluster" : null));
```